cpp/src/sampling/renumber_sampled_edgelist_impl.cuh

```diff
@@ -0,0 +1,723 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <prims/kv_store.cuh>
+
+#include <cugraph/utilities/device_functors.cuh>
+#include <cugraph/utilities/misc_utils.cuh>
+
+#include <raft/core/handle.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cub/cub.cuh>
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/distance.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/merge.h>
+#include <thrust/set_operations.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+#include <optional>
+
+namespace cugraph {
+
+namespace {
+
+template <typename vertex_t, typename label_index_t>
+std::tuple<rmm::device_uvector<vertex_t>, std::optional<rmm::device_uvector<label_index_t>>>
+compute_renumber_map(raft::handle_t const& handle,
+                     raft::device_span<vertex_t const> edgelist_srcs,
+                     std::optional<raft::device_span<int32_t const>> edgelist_hops,
+                     raft::device_span<vertex_t const> edgelist_dsts,
+                     std::optional<raft::device_span<size_t const>> label_offsets)
+{
+  auto approx_edges_to_sort_per_iteration =
+    static_cast<size_t>(handle.get_device_properties().multiProcessorCount) *
+    (1 << 20) /* tuning parameter */;  // for segmented sort
+
+  std::optional<rmm::device_uvector<label_index_t>> edgelist_label_indices{std::nullopt};
+  if (label_offsets) {
+    edgelist_label_indices =
```

Comment:
Is this the same logical function as https://github.com/rapidsai/cugraph/blob/cab836896aeba2c5549c57ebfd4818e883c17b5c/cpp/src/structure/detail/structure_utils.cuh#L87?

If so, perhaps this should call that function.  If this implementation is better then perhaps we should swap the implementation to the better one.

Reply:
What about moving expand_sparse_offsets (https://github.com/rapidsai/cugraph/blob/cab836896aeba2c5549c57ebfd4818e883c17b5c/cpp/src/structure/detail/structure_utils.cuh#L87) to misc_utils.cuh (https://github.com/rapidsai/cugraph/blob/branch-23.08/cpp/include/cugraph/utilities/misc_utils.cuh).

We are using this function multiple places outside `src/structure` (including few files under the `test` directory).

Reply:
That sounds like a great idea.  It's very reasonable to be part of the public API.

Reply:
OK, I addressed your review comments.